### PR TITLE
CFE-4281: LMDB testbed

### PR DESCRIPTION
**Pull request by vpodzime**
Remaining things:
- [x] `gettid()` using `syscall()`
- [x] Python bindings
- [x] transaction committing
- [x] elementary testing

Merge together:
https://github.com/cfengine/core/pull/5438
https://github.com/cfengine/buildscripts/pull/1367

### CODE REVIEW

**Comment by vpodzime** [comment]
@olehermanse @larsewi when reviewing, please keep in mind that this is code that **should not** be used by our users and that is not even be part of our release packages. It's for internal usage only. At least for now, we can definitely change this in the future (part of `cf-check`, maybe?). The same applies to support of exotics.

File: libpromises/dbm_test_api.c [hunk 1]
```diff
@@ -0,0 +1,724 @@
...
+}
+
+
+void DoRandomIterations(dbid db_id,
+                        long min_interval_ms, long max_interval_ms,
+                        bool *terminate)
+{
+    assert(terminate != NULL);
+
+    InitializeRNG((long)gettid());
```
Comment (larsewi) [inline libpromises/dbm_test_api.c:370]: No big deal, but:

> Type casts should be separated with one space from the variable
- [CONTRIBUTING.md](https://github.com/cfengine/core/blob/master/CONTRIBUTING.md#formatting--whitespace)

File: python/README [hunk 2]
```diff
@@ -0,0 +1,10 @@
+This directory contains Python packages and modules for working with CFEngine
+and its libraries.
+
+If you want to use one of these, PYTHONPATH and LD_LIBRARY_PATH variables can
+make things easier when used like this:
+
+  PYTHONPATH=/var/cfengine/lib/python LD_LIBRARY_PATH=/var/cfengine/lib python3
+
+That will make sure that the bindings are found by Python and that CFEngine's
+libraries are found by dlopen() (used by ctypes.CDLL() etc.).
```
Comment (larsewi) [inline python/README:10]:
```suggestion
libraries are found by dlopen() (used by ctypes.CDLL() etc.).

```

**Review by larsewi — DISMISSED** [review]
Cool 🚀

File: libpromises/dbm_test_api.c [hunk 3]
```diff
@@ -0,0 +1,724 @@
...
+static void InitializeRNG(long int seed)
+{
+    srand48_r(seed, &rng_data);
+}
+
+static long GetRandomNumber(long limit)
+{
+    long rnd_val;
+    lrand48_r(&rng_data, &rnd_val); /* generates a value in the [0, 2^31) interval */
+    const long random_max = ((0x80000000 - 1) / limit) * limit;
```
[the remark this answers is not in the document]
Comment (olehermanse) [reply libpromises/dbm_test_api.c:84]: @larsewi It's _almost_ the same, but this is integer division. It's to get rid of the small part at the end which is not a multiple of your limit, to eliminate modulo bias, as described here:

https://stackoverflow.com/questions/10984974/why-do-people-say-there-is-modulo-bias-when-using-a-random-number-generator

The code could probably use a comment / link about this.

File: libpromises/dbm_test_api.c [hunk 4]
```diff
@@ -0,0 +1,724 @@
...
+}
+
+static __thread struct drand48_data rng_data;
+
+static void InitializeRNG(long int seed)
+{
+    srand48_r(seed, &rng_data);
+}
+
+static long GetRandomNumber(long limit)
```
[the remark this answers is not in the document]
Comment (olehermanse) [reply libpromises/dbm_test_api.c:80]: See other comment about modulo bias; https://stackoverflow.com/questions/10984974/why-do-people-say-there-is-modulo-bias-when-using-a-random-number-generator

File: libpromises/dbm_test_api.c [hunk 5]
```diff
@@ -0,0 +1,724 @@
...
+    int keys_refresh_s; long min_interval_ms; long max_interval_ms;
+    bool *terminate;
+};
+static void *DoRandomReadsRoutine(void *data)
+{
+    struct ReadParams *params = data;
+    DoRandomReads(params->db_id,
+                  params->keys_refresh_s, params->min_interval_ms, params->max_interval_ms,
+                  params->terminate);
+    return NULL;
```
Comment (olehermanse) [inline libpromises/dbm_test_api.c:222]: Not imediately clear why this function shouldn't have a return type;

```suggestion
    // Always return NULL, 
    // pthread_create wants a function which returns (void *)
    return NULL;
```

File: python/Makefile.am [hunk 6]
```diff
@@ -0,0 +1,33 @@
+#
+#  Copyright 2021 Northern.tech AS
```
Comment (olehermanse) [inline python/Makefile.am:2]:
```suggestion
#  Copyright 2024 Northern.tech AS
```

File: libpromises/dbm_test_api.h [hunk 7]
```diff
@@ -0,0 +1,57 @@
+/*
+  Copyright 2023 Northern.tech AS
```
Comment (olehermanse) [inline libpromises/dbm_test_api.h:2]:
```suggestion
  Copyright 2024 Northern.tech AS
```

**Review by olehermanse — DISMISSED** [review]
Cool, thanks for working on this! Left some smaller comments, but didn't find anything serious, so looks good to me.

File: libpromises/dbm_test_api.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (larsewi) [reply libpromises/dbm_test_api.c:80]: Thanks, this makes sense 🚀

File: libpromises/dbm_test_api.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (vpodzime) [reply libpromises/dbm_test_api.c:84]: Yup. It's a bit of an overkill for this use-case, but who knows where this code ends up in the future :)

File: libpromises/dbm_test_api.c [hunk 8]
```diff
@@ -0,0 +1,724 @@
...
+    {
+        /* got a bad value past the greatest multiple of the limit interval,
+         * retry */
+        lrand48_r(&rng_data, &rnd_val);
+    }
+    return rnd_val % limit;
+}
+
+
+static Seq *GetDBKeys(DBHandle *db)
```
[the remark this answers is not in the document]
Comment (vpodzime) [reply libpromises/dbm_test_api.c:95]: I'll go through the code, but I intentionally put 0-2 blank lines between things to group/separate them together/from each other a bit more.

File: libpromises/dbm_test_api.c [hunk 9]
```diff
@@ -0,0 +1,724 @@
...
+            success = NextDB(cur, &key, &key_size, &val, &val_size);
+        }
+        DeleteDBCursor(cur);
+        CloseDB(db);
+    }
+}
+struct IterParams {
+    dbid db_id;
+    long min_interval_ms; long max_interval_ms;
+    bool *terminate;
```
[the remark this answers is not in the document]
Comment (vpodzime) [reply libpromises/dbm_test_api.c:413]: Opinion noticed. :)

**Pull request merged**